From e7f9c0a955057820c1b26e13593acb1b41b06ab9 Mon Sep 17 00:00:00 2001
From: Vinicius Costa Gomes <vinicius.gomes@intel.com>
Date: Wed, 23 Jun 2021 23:01:48 -0700
Subject: [PATCH 112/122] igc: Correct the launchtime offset

According to section 7.5.2.6 of the datasheet, the launchtime offset
can be corrected.

Preliminary measurements indicate that this correction is around
1500ns. Proper measurements need to be made and the corrections need
to consider the different link speeds.

Signed-off-by: Vinicius Costa Gomes <vinicius.gomes@intel.com>
Signed-off-by: Muhammad Husaini Zulkifli <muhammad.husaini.zulkifli@intel.com>
---
 drivers/net/ethernet/intel/igc/igc_regs.h | 2 ++
 drivers/net/ethernet/intel/igc/igc_tsn.c  | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/drivers/net/ethernet/intel/igc/igc_regs.h b/drivers/net/ethernet/intel/igc/igc_regs.h
index 3d27998e6f23..3d718b14ef30 100644
--- a/drivers/net/ethernet/intel/igc/igc_regs.h
+++ b/drivers/net/ethernet/intel/igc/igc_regs.h
@@ -217,6 +217,8 @@
 #define IGC_QBVCYCLET		0x331C
 #define IGC_QBVCYCLET_S		0x3320
 
+#define IGC_GTXOFFSET		0x3310
+
 #define IGC_STQT(_n)		(0x3324 + 0x4 * (_n))
 #define IGC_ENDQT(_n)		(0x3334 + 0x4 * (_n))
 #define IGC_DTXMXPKTSZ		0x355C
diff --git a/drivers/net/ethernet/intel/igc/igc_tsn.c b/drivers/net/ethernet/intel/igc/igc_tsn.c
index 1dde53f60c8a..9708d625e587 100644
--- a/drivers/net/ethernet/intel/igc/igc_tsn.c
+++ b/drivers/net/ethernet/intel/igc/igc_tsn.c
@@ -78,6 +78,8 @@ static int igc_tsn_enable_offload(struct igc_adapter *adapter)
 	u32 frag_size_mult;
 	int i;
 
+	wr32(IGC_GTXOFFSET, 1500);
+
 	wr32(IGC_TSAUXC, 0);
 	wr32(IGC_DTXMXPKTSZ, IGC_DTXMXPKTSZ_TSN);
 	wr32(IGC_TXPBS, IGC_TXPBSIZE_TSN);
-- 
2.17.1

